File: frontend/types/activity.ts
export interface GarminSplit {
  noOfSplits: number;
  index?: number; // computed split index (1-based) for per-km splits
  splitType?: string;
  // cumulative values (as stored by backend)
  cumulative_time_s?: number; // seconds
  cumulative_distance_m?: number; // meters
  // per-split values
  split_time_s?: number; // seconds for this split (delta)
  split_distance_m?: number; // meters for this split (delta)
  pace_s_per_km?: number; // seconds per km

  // original Garmin fields (kept for compatibility)
  duration?: number; // seconds (sometimes cumulative)
  distance?: number; // meters (sometimes cumulative)
  averageSpeed?: number; // m/s
  maxSpeed?: number; // m/s
  totalAscent?: number;
  elevationLoss?: number;
  maxElevationGain?: number;
  averageElevationGain?: number;
  numFalls?: number;
  numClimbSends?: number;
}

export interface Activity {
  id: number;
  garmin_id: string;
  name: string | null;
  activity_type: string;
  activity_type_key: string | null;
  start_time: string;
  start_time_local: string;
  duration_seconds: number | null;
  elapsed_duration_seconds: number | null;
  moving_duration_seconds: number | null;
  distance_meters: number | null;
  average_speed: number | null;
  max_speed: number | null;
  average_pace_seconds: number | null;
  average_heart_rate: number | null;
  max_heart_rate: number | null;
  calories: number | null;
  elevation_gain: number | null;
  elevation_loss: number | null;

  // Running metrics
  average_cadence: number | null;
  average_running_cadence: number | null;
  max_running_cadence: number | null;
  average_stride_length: number | null;
  vertical_oscillation: number | null;
  ground_contact_time: number | null;
  vo2_max: number | null;
  lap_count: number | null;
  has_splits: boolean | null;
  has_intensity_intervals: boolean | null;
  fastest_split_1000: number | null;
  fastest_split_1609: number | null;

  // Swimming metrics
  pool_length: number | null;
  average_swolf: number | null;
  average_swim_cadence: number | null;
  stroke_type: string | null;
  active_lengths: number | null;
  strokes: number | null;
  average_strokes: number | null;
  water_estimated: number | null;

  // Strength / resistance training (support both camelCase from backend JSON and snake_case)
  summarized_exercise_sets?: {
    category?: string | null;
    reps?: number | null;
    volume?: number | null;
    duration?: number | null;
    sets?: number | null;
    maxWeight?: number | null;
  }[] | null;
  total_sets?: number | null;
  active_sets?: number | null;
  total_reps?: number | null;
  max_weight?: number | null;

  // Splits & Laps
  splits: GarminSplit[] | null;
  laps: any | null;
}

File: backend/app/schemas/activity.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Any


class ActivityBase(BaseModel):
    garmin_id: str
    name: str | None = None
    activity_type: str
    activity_type_key: str | None = None
    start_time: datetime
    start_time_local: datetime | None = None
    timezone: str | None = None
    duration_seconds: float | None = None
    elapsed_duration_seconds: float | None = None
    moving_duration_seconds: float | None = None
    distance_meters: float | None = None
    average_speed: float | None = None
    max_speed: float | None = None
    average_pace_seconds: float | None = None
    average_heart_rate: int | None = None
    max_heart_rate: int | None = None
    calories: int | None = None
    elevation_gain: float | None = None
    elevation_loss: float | None = None
    average_cadence: int | None = None
    average_running_cadence: float | None = None
    max_running_cadence: float | None = None
    average_stride_length: float | None = None
    vertical_oscillation: float | None = None
    ground_contact_time: float | None = None
    lap_count: int | None = None
    has_splits: bool | None = None
    has_intensity_intervals: bool | None = None
    fastest_split_1000: float | None = None
    fastest_split_1609: float | None = None
    pool_length: float | None = None
    average_swolf: float | None = None
    average_swim_cadence: float | None = None
    stroke_type: str | None = None
    active_lengths: int | None = None
    strokes: float | None = None
    average_strokes: float | None = None
    water_estimated: float | None = None
    start_latitude: float | None = None
    start_longitude: float | None = None
    has_gps: bool = False
    training_effect: float | None = None
    anaerobic_training_effect: float | None = None
    vo2_max: float | None = None


class ActivityResponse(ActivityBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    polyline: str | None = None
    splits: Any | None = None
    laps: Any | None = None
    created_at: datetime
    updated_at: datetime
    synced_at: datetime | None = None

    # Computed helpers
    @property
    def pace_per_km(self) -> str | None:
        if self.average_pace_seconds:
            minutes = int(self.average_pace_seconds // 60)
            seconds = int(self.average_pace_seconds % 60)
            return f"{minutes}:{seconds:02d}"
        return None

    @property
    def distance_km(self) -> float | None:
        if self.distance_meters:
            return round(self.distance_meters / 1000, 2)
        return None

    @property
    def duration_formatted(self) -> str | None:
        if self.duration_seconds:
            hours = int(self.duration_seconds // 3600)
            minutes = int((self.duration_seconds % 3600) // 60)
            seconds = int(self.duration_seconds % 60)
            if hours > 0:
                return f"{hours}:{minutes:02d}:{seconds:02d}"
            return f"{minutes}:{seconds:02d}"
        return None


class ActivityListResponse(BaseModel):
    activities: list[ActivityResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class ActivityFilters(BaseModel):
    activity_type: str | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    search: str | None = None
    page: int = 1
    page_size: int = 20
